### PR TITLE
serialize dependencies into protobuf

### PR DESCRIPTION
**Pull request by changpingc**
We'd like to ship dependencies out of thunder server process, which allows us to restore a subscription and avoid expensive initial computation when a client reconnects and resumes prior subscriptions. This also enables external dependency invalidator that can make thunder server stateless.

### CODE REVIEW

**Comment by berfarah** [comment]
@changpingc Do you mind putting the reverts in a separate PR? 😬

**Comment by changpingc** [comment]
revert landed separately in #167.

File: sqlgen/reflect.go [hunk 1]
```diff
@@ -815,3 +815,42 @@ func (t *Table) extractRow(row interface{}) Filter {
...
+	case string:
+		return string(v)
+	case []byte:
+		return []byte(v)
+	case float32:
+		return float64(v)
+	case float64:
+		return float64(v)
+	case time.Time:
+		return time.Time(v)
```
Comment (berfarah) [inline sqlgen/reflect.go]: There are some types here that we're doing nothing with. Do we need to handle them?

File: livesql/marshal.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/samsarahq/thunder/sqlgen"
+	"github.com/samsarahq/thunder/thunderpb"
+)
+
+func valueToField(value interface{}) (*thunderpb.Field, error) {
+	switch column := sqlgen.MakeCanonical(value).(type) {
```
Comment (berfarah) [inline livesql/marshal.go]: What happens for custom types?

File: reactive/rerunner.go [hunk 3]
```diff
@@ -131,14 +133,43 @@ func (r *Resource) Cleanup(f func()) {
...
-func AddDependency(ctx context.Context, r *Resource) {
+type dependencySetKey struct{}
+
+type dependencySet struct {
+	mu           sync.Mutex
+	dependencies []Serializable
+}
+
+type Serializable interface {
+	proto.Message
```
[the remark this answers is not in the document]
Comment (changpingc) [reply reactive/rerunner.go]: sure it can turn this into Marshal(), Unmarshal() probably

File: sqlgen/reflect.go (hunk 1 above)
Comment (changpingc) [reply sqlgen/reflect.go]: this handles all driver.Value.

File: livesql/marshal.go (hunk 2 above)
Comment (changpingc) [reply livesql/marshal.go]: Talked offline. We are going to support it by storing driver.Value. See last commit!

File: reactive/rerunner.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (changpingc) [reply reactive/rerunner.go]: done

File: sqlgen/reflect.go (hunk 1 above)
Comment (berfarah) [reply sqlgen/reflect.go]: This handles more than `driver.Value` - for example, `uint32` is not a `driver.Value`, only `int64` is

File: reactive/rerunner.go [hunk 4]
```diff
@@ -131,14 +131,46 @@ func (r *Resource) Cleanup(f func()) {
...
 		r.node.addOut(&node{released: true})
 		return
 	}
 
 	computation := ctx.Value(computationKey{}).(*computation)
 	r.node.addOut(&computation.node)
+
+	if serializable != nil {
+		depSet, ok := ctx.Value(dependencySetKey{}).(*dependencySet)
+		if ok && depSet != nil {
```
[the remark this answers is not in the document]
Comment (changpingc) [reply reactive/rerunner.go]: done

File: reactive/rerunner.go [hunk 5]
```diff
@@ -131,14 +131,46 @@ func (r *Resource) Cleanup(f func()) {
...
+type dependencySetKey struct{}
+
+type dependencySet struct {
+	mu           sync.Mutex
+	dependencies []Serializable
+}
+
+type Serializable interface {
+	// proto.Marshaler and proto.Unmarshaler.
+	Marshal() ([]byte, error)
```
[the remark this answers is not in the document]
Comment (changpingc) [reply reactive/rerunner.go]: I don't particularly like that..

File: sqlgen/reflect.go (hunk 1 above)
Comment (changpingc) [reply sqlgen/reflect.go]: opened a separate PR https://github.com/samsarahq/thunder/pull/170

**Comment by changpingc** [comment]
@berfarah I think this is ready to land! It should not introduce any additional error in existing API.

**Comment by berfarah** [comment]
@changpingc given the amount of changes here, do you think a second pair of eyes might be useful?

File: reactive/rerunner.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (berfarah) [reply reactive/rerunner.go]: Yeah but would also be a built-in interface. I'm ok with not wrapping.

File: livesql/marshal_test.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+				"id": int32(1),
+			},
+			unmarshaled: sqlgen.Filter{
+				"id": int64(1),
+			},
+		},
+		{
+			name: "id int64 ptr to int64",
+			filter: sqlgen.Filter{
+				"id": &one,
```
Comment (berfarah) [inline livesql/marshal_test.go]: nil test case?

File: thunderpb/dependency.proto [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+  Bool = 2;
+  Int = 3;
+  Uint = 4;
+  String = 5;
+  Bytes = 6;
+  Float64 = 7;
+  Time = 8;
+}
+
+message Field {
```
Comment (willhug) [inline thunderpb/dependency.proto]: ahh, think I found what's causing a lot of cloudserver OOMs, I think this field is quite large and taking up much more memory than usual whenever we create a dependency.

Have we thought of using the proto oneof annotation here?  Not sure if it actually reduces the allocation in code, just a thought.

Comment (willhug) [reply thunderpb/dependency.proto]: here's one of our heap profiles: https://s3.internal.samsara.com/s3/samsara-prod-profile/2018-10-02/cloudserver/1438-prod-23996c8a-f0f5-47be-960a-d96d639a2f26-2wHXqesa-heap.pprof.svg

**Pull request merged**